SNS: rename config name settings

jsilhan opened the pull request:
in `/pymess/config.py` is `SMS_SNS_CONFIG` name so I believe it should be consistent with `SMS_OPERATOR_CONFIG`, etc.

jsilhan commented on pymess/backend/sms/sns.py:
```diff
@@ -23,9 +23,9 @@ def _get_sns_client(self):
-                region_name=settings.SNS.AWS_SMS_REGION,
+                aws_access_key_id=settings.SMS_SNS_CONFIG.AWS_ACCESS_KEY_ID,
+                aws_secret_access_key=settings.SMS_SNS_CONFIG.AWS_SECRET_ACCESS_KEY,
+                region_name=settings.SMS_SNS_CONFIG.AWS_REGION,
```
`AWS_SMS_REGION` -> `AWS_REGION`

jsilhan commented on pymess/backend/sms/__init__.py:
```diff
@@ -101,8 +99,21 @@ def bulk_check_sms_states(self):
+                state=self.model.STATE.ERROR, error=_('timeouted')
+            )
+
+    def _is_recipient_phone_number_allowed(self, message):
```
so it can be safely tested on the server with other valid phone numbers

jsilhan replied:
I wanted to make one Generic config option for all SMS backends but in code it looks complicated :(

rubickcz replied:
I see you are using a generic option, or?

matllubos replied:
maybe better solution is have dynamic backend configuration and for spicific number will be selected specific backend

jsilhan commented:
@rubickcz @matllubos I have fixed a few issues in SNS backend and adde a new `SMS_TESTING_PHONES` option. Please take a look again.

rubickcz commented on pymess/backend/sms/__init__.py:
```diff
@@ -101,8 +99,21 @@ def bulk_check_sms_states(self):
+            )
+
+    def _is_recipient_phone_number_allowed(self, message):
+        recipient = message.recipient
```
unnecessary variable

rubickcz commented on pymess/backend/sms/ats_sms_operator.py:
```diff
@@ -230,30 +230,32 @@ def _update_sms_states_from_response(self, messages, parsed_response, is_sending
 
     def publish_messages(self, messages):
+        messages = (message for message in messages if self._is_recipient_phone_number_allowed(message))
         self._send_requests(messages, request_type=self.REQUEST_TYPES.SMS, is_sending=True, sent_at=timezone.now())
```
I don't understand why this method is overriden anyway. In parent, it just calls `publish_message` for each message. Here, it is doing the same thing as `publish_message` does, so what's the point.

jsilhan replied:
this `publish_messages` method sends all messages in one request. Otherwise it would do one request per message.

rubickcz commented on pymess/backend/sms/ats_sms_operator.py:
```diff
@@ -230,30 +230,32 @@ def _update_sms_states_from_response(self, messages, parsed_response, is_sending
-            )
-            # Do not re-raise caught exception. Re-raise exception causes transaction rollback (lost of information
-            # about exception).
+        if self._is_recipient_phone_number_allowed(message):
```
I think you should do this check in parent, otherwise in new backends, it will be easy to forget calling this. Something like:
```
def publish_message(self, message):
    if self._can_publish_message(message):
        self._publish_message(message)

def _publish_message(self, message):
    raise NotImplementedError
```

matllubos replied:
better will be solution with dynamic backends

rubickcz replied:
umm, ok, but that will take some time to rewrite

matllubos commented on pymess/backend/sms/__init__.py:
```diff
@@ -101,8 +99,21 @@ def bulk_check_sms_states(self):
+    def _is_recipient_phone_number_allowed(self, message):
+        recipient = message.recipient
+        testing_phones = settings.SMS_TESTING_PHONES
+        if testing_phones and recipient not in testing_phones:
```
this is not nice solution

jsilhan commented on docs/dialer.rst:
```diff
@@ -205,10 +205,6 @@ If you want to write your own Pymess dialer backend you must create class that i
 
     This method should send dialer message (obtained from the input argument) and update its state. This method must be overridden in the custom backend.
 
-  .. method:: publish_messages(messages)
```
This is the main issue I encountered with dynamic backends. A lot of methods should be actually called outside of certain backend and inside the function the backend would be chosen from backend type and recipient. You will see what I mean.
Also this would change lib API.
Methods `create_message`, `publish_messages` and `bulk_send` needed to be refactored (they are called from our core project)
Once we find the proper solution I will change the docs accordingly.

jsilhan commented:
@matllubos @rubickcz I rewrote it, please take a look

matllubos commented on docs/sms.rst:
```diff
@@ -232,13 +236,6 @@ If you want to write your own Pymess SMS backend, you must create class that inh
 
     This method should send SMS message (obtained from the input argument) and update its state. This method must be overridden in the custom backend.
 
-  .. method:: publish_messages(messages)
```
be carefull this function is used in core

matllubos commented on pymess/backend/__init__.py:
```diff
@@ -192,25 +191,6 @@ def send(self, recipient, content, related_objects=None, tag=None, template=None
         return message
 
-    @transaction.atomic
-    def bulk_send(self, recipients, content, related_objects=None, tag=None, template=None, **kwargs):
```
this maybe can work bud inside recipients can be numbers/e-mails for this specific backend

jsilhan replied:
it's inside controller now

matllubos commented on pymess/backend/sms/__init__.py:
```diff
@@ -1,18 +1,16 @@
-from pymess.backend import BaseBackend, send_template as _send_template, send as _send
-from pymess.config import settings, get_sms_template_model, get_sms_sender
+from django.utils import timezone
+from django.utils.translation import ugettext as _
```
weird sort

matllubos commented on pymess/backend/sms/__init__.py:
```diff
@@ -141,6 +116,36 @@ def send(recipient, content, related_objects=None, tag=None, **kwargs):
+    """
+    Method that find messages that is not in the final state and updates its states.
+    """
+    for backend in {get_sms_sender(), get_sms_sender(is_testing=True)}:
```
this will not work if you change backend for the old messages

matllubos replied:
I think you must get backends from SMS messages

matllubos commented on pymess/functions.py:
```diff
@@ -0,0 +1,55 @@
+    :param template: template object from which content of the message was created
+    :param kwargs: extra attributes that will be saved with the message
+    """
+    return backend_fnc(recipient=recipient).create_message(recipient, content, related_objects, tag, template, **kwargs)
```
why this shoud do backend inside

jsilhan replied:
In general this should work as expected without finding the correct backend assuming for each backend types is the same implementation. But to make it 100% correct it should call  `create_message` for the right backend IMO - `create_message` is a custom method of concrete backend and can be overridden.

jsilhan replied:
with controller separation this is not needed.

jsilhan replied:
refactored

matllubos commented on pymess/functions.py:
```diff
@@ -0,0 +1,55 @@
+        except KeyError:
+            backends_messages_map[backend] = [message]
+    for backend, messages in backends_messages_map:
+        backend._publish_messages(messages)
```
you shoud not call protected method. publish_messages must be public

jsilhan replied:
I wanted to make it clear we don;r want this method to be called from the outside and use this function instead.

jsilhan replied:
It's public now. I just changed the name to `bulk_send_messages` in controller and kept also `publish_messages` on backend (to differentiate these two)

matllubos commented on pymess/functions.py:
```diff
@@ -0,0 +1,55 @@
+    :param backend_fnc: get backend function which will be evaluated based on recipient
+    :param messages: list of messages
+    """
+    backends_messages_map = {}
```
use defaultdict

jsilhan replied:
fixed

matllubos commented on pymess/management/commands/send_messages_batch.py:
```diff
@@ -69,17 +62,15 @@ def _print_result(self, title, message_pks):
-        except DatabaseError:
-            self.stdout.write('messages are already locked')
-
+        for sender in {self.senders[type](), self.senders[type](is_testing=True)}:
```
so message is created without backend and backend is used only for sending message? 

So:
* batch_size cannot be defined in the backend (more backends can be used during sending messages)
* backend cannot manage if batch sending is turned on or off (all the backends must have same settings)

matllubos replied:
you must get backend per recipient here too

matllubos replied:
For my opinion there must be controller for every type of message (email, sms, dialer,...) Controler will manage creating message, batch sending, etc. And backend only manage sending the message to the service. So you must split backend to two classes now

jsilhan replied:
ok, I will refactor it

jsilhan replied:
I have refactored it. Only `is_turned_on_batch_sending` is still in the backend (because it's being accessed when the message is being send within the backend. I didn't want to do backend -> controller relation, there's only controller -> backend

matllubos commented on pymess/config.py:
```diff
@@ -166,14 +172,19 @@ def get_email_template_model():
     """
-    return import_string(settings.SMS_SENDER_BACKEND)()
+    testing_phones = settings.SMS_TESTING_PHONES
+    if is_testing == True or is_testing is None and (testing_phones and recipient in testing_phones):
```
the is_testing is weird this shoud be in the configuration only

jsilhan replied:
removed

jsilhan commented on pymess/backend/dialer/__init__.py:
```diff
@@ -14,73 +15,55 @@
+        return self.model.STATE.WAITING
 
-    def create_message(self, recipient, content, related_objects, tag, template, is_autodialer=True, **kwargs):
+    def create_message(self, recipient, content=None, related_objects=None, tag=None, template=None,
```
copied from daktela backend

jsilhan commented on pymess/backend/sms/sns.py:
```diff
@@ -15,16 +15,20 @@ class SNSSMSBackend(SMSBackend):
 
     sns_client = None
 
+    def __init__(self):
```
SNS changes below:

jsilhan commented on pymess/management/commands/send_messages_batch.py:
```diff
@@ -69,17 +65,15 @@ def _print_result(self, title, message_pks):
-
-        if not sender.is_turned_on_batch_sending():
-            raise CommandError('Batch sending is turned off')
+        controller = self.controllers[type]
```
the problematic `is_turned_on_batch_sending` part... It's needed to send messages from backend

jsilhan replied:
refactored

jsilhan commented:
@matllubos @rubickcz comments have been addressed. Please review again.

matllubos commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,130 @@
 
-    def __init__(self):
-        assert self.model is not None, _l('self.model cannot be None')
+    class BackendNotConfiguredError(Exception):
```
dej tu exception ven

matllubos commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,130 @@
+        for backend, recipients_allowed in self.get_backends_dict().items():
+            if not recipients_allowed or recipient in recipients_allowed:
+                return backend
+        raise self.BackendNotConfiguredError
```
spíš bych to asi i přejmenovat. Na BackendNotFound

matllubos replied:
a přidal tomu strin,mže pro tohoto příjemce nebyl backend nalezen

matllubos commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,130 @@
+            message.related_objects.create_from_related_objects(*related_objects)
+        return message
+
+    def _get_backend_messages_map(self, messages):
```
tohle ti trošku bude kazit tu prioritu co děláš v jiném tasku

matllubos commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,130 @@
+        for backend, messages_for_backend in self._get_backend_messages_map(messages):
+            backend.publish_messages(messages_for_backend)
+
+    @transaction.atomic
```
pozor na to atomic, když ti padne jedna zpráva tak se ti nesmí udělat už rollback na ty předchozí (ten atomic už tam byl?)

matllubos commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,130 @@
+        messages = [
+            self.create_message(recipient, content, related_objects, tag, template, **kwargs)
+            for recipient in recipients
+        ]
```
vlastně by atomic měl být jen při vytvoření zpráv

matllubos commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,130 @@
+            self.create_message(recipient, content, related_objects, tag, template, **kwargs)
+            for recipient in recipients
+        ]
+        self.bulk_send_messages(messages)
```
a pak při odeslání per jedna zpráva, případně když jsme si jisti, že backend nevrátí error tak už bych tam ani atomic nedával

matllubos commented on pymess/backend/__init__.py:
```diff
@@ -71,10 +174,10 @@ def _update_message_after_sending_error(self, message, extra_sender_data=None, s
-                        number_of_send_attempts > self.get_batch_max_number_of_send_attempts()
-                        or not self.get_retry_sending()
-                ) else self.model.STATE.ERROR_RETRY
+                message.__class__.STATE.ERROR if (
```
přoč je tam nutné to .__class__ ?

matllubos commented on pymess/backend/dialer/__init__.py:
```diff
@@ -14,73 +15,56 @@
-        return settings.DIALER_BATCH_SENDING
+    def get_backends_dict(self):
+        return OrderedDict([
+            (get_dialer_sender(), []),
```
tohle by se mělo tahat ze settings ne?

matllubos commented on pymess/backend/dialer/__init__.py:
```diff
@@ -89,11 +73,38 @@ def bulk_check_dialer_status(self):
+    is used for automatic call to selected phone number.
+    """
+
+    def is_turned_on_batch_sending(self):
```
neměl by tento konfig být v controlleru?

matllubos commented on pymess/backend/emails/__init__.py:
```diff
@@ -86,16 +85,29 @@ def send(self, sender, recipient, subject, content, sender_name=None, related_ob
+    is used for sending messages.
+    """
+
+    def is_turned_on_batch_sending(self):
```
opět tohle má řídit controller, protože je jen jedno settings..

matllubos commented on pymess/backend/emails/dummy.py:
```diff
@@ -10,7 +9,7 @@ class DummyEmailBackend(EmailBackend):
 
     def publish_message(self, message):
-        self._update_message_after_sending(message, state=EmailMessage.STATE.DEBUG)
+        self._update_message_after_sending(message, state=message.__class__.STATE.DEBUG)
```
proč __class__

matllubos commented on pymess/backend/sms/__init__.py:
```diff
@@ -1,46 +1,46 @@
 
+    def get_backends_dict(self):
+        return OrderedDict([
+            (get_sms_testing_sender(), settings.SMS_TESTING_PHONES),
```
opět tohle se má nastavovat v settings. Cesty k backendům per čísla. Takto to není univerzální

jsilhan replied:
I have renamed the default sender backend config then to `_1`

jsilhan commented:
@matllubos comments addressed. Please review again.

matllubos commented on pymess/backend/dialer/__init__.py:
```diff
@@ -89,11 +72,38 @@ def bulk_check_dialer_status(self):
+    def get_batch_max_number_of_send_attempts(self):
+        return settings.DIALER_BATCH_MAX_NUMBER_OF_SEND_ATTEMPTS
+
+    def get_batch_max_seconds_to_send(self):
```
používá se to v publish_or_retry_message to by teoreticky také mohlo být v controlleru, nebo máš nějaký specifický důvod, proč je to zde?

jsilhan replied:
right, I moved `publish_or_retry_message` with `get_batch_max_seconds_to_send` into Controler.

I will keep there: `get_retry_sending` and `get_batch_max_number_of_send_attempts` (it is inside `_update_message_after_sending_error`). `_update_dialer_states` is specific to backend so I would keep it here as well.

matllubos commented on pymess/config.py:
```diff
@@ -22,10 +25,14 @@
     'SMS_USE_ACCENT': False,
     'SMS_DEFAULT_PHONE_CODE': None,
-    'SMS_SENDER_BACKEND': 'pymess.backend.sms.dummy.DummySMSBackend',
+    'SMS_SENDER_BACKEND_1': 'pymess.backend.sms.dummy.DummySMSBackend',
```
tohle určitě ne, to musíme nastavovat v listu

jsilhan commented on docs/push.rst:
```diff
@@ -188,7 +188,7 @@ Backend is a class that is used for sending messages. Every backend must provide
   Configuration of attributes according to push notification operator documentation::
 
-    PYMESS_PUSH_NOTIFICATION_ONESIGNAL = {
+    config = {
```
it's configured within `*_BACKENDS` settings now (see Pr in core)

jsilhan commented on pymess/backend/dialer/daktela.py:
```diff
@@ -108,19 +123,10 @@ def publish_message(self, message):
-                message,
-                error=_('Not allowed number to dial. Allowed numbers are: {}').format(', '.join(testing_phones)),
-                state=DialerMessage.STATE.DEBUG,
-                is_final_state=True
```
`is_final_state=True` is set in DummyDialerBackend now

jsilhan commented:
@rubickcz @matllubos I have fixed another issues and refactored settings to Routers. please review again. (I recommend you to start with high level PR in core)

matllubos commented on pymess/config.py:
```diff
@@ -1,31 +1,25 @@
 
 from attrdict import AttrDict
 
+DEFAULT_SENDER_BACKEND_NAME = 'default'
```
add this to the settings please

matllubos replied:
per contoller (sms, e-mail, push,...)

jsilhan replied:
added

matllubos commented:
I like this pr

rubickcz commented on pymess/config.py:
```diff
@@ -121,23 +95,31 @@
+        try:
+            default_value = DEFAULTS[attr]
+        except KeyError:
+            if re.search(r'_\d+$', attr):
```
You don't need this anymore, or?

jsilhan replied:
it's leftover from the previous change. It's not needed now. Removed. Thanks.

rubickcz commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,156 @@
+
+    config = AttrDict()
+
+    def __init__(self, configurations=None):
```
Why no just `config`?

jsilhan replied:
changed

jsilhan commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,156 @@
-        assert self.model is not None, _l('self.model cannot be None')
+        self._loaded_backends = {}
+
+    def get_backend(self, recipient):
```
huh, And at first it was just about trying AWS SNS... :)
We insert into message additional params when it's created.
```
return super().create_message(
                recipient=recipient,
                content=content,
                related_objects=related_objects,
                tag=tag,
                template=template,
                state=self.get_initial_sms_state(recipient),
                extra_data=kwargs,
                **self.get_backend(recipient).get_extra_message_kwargs()
            )
```

jsilhan replied:
so still we would nee to define which parameters can be used by router and which could be added by extra_message_kwargs

matllubos replied:
maybe now keep this solution in the future we can add whole message to the router...

rubickcz replied:
agreed.

jsilhan commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,156 @@
+            messages = [
+                self.create_message(recipient, content, related_objects, tag, template, **kwargs)
+                for recipient in recipients
+            ]
```
this is called from admin so I would say we expect all or nothing... for the creation part.
The whole method was in atomic but we want to actually at least know which messages were sent (when error occurred)

jsilhan commented on pymess/backend/routers.py:
```diff
@@ -0,0 +1,13 @@
+class BaseRouter:
+
+    def get_backend(self, recipient):
```
right, fixed.

jsilhan commented on pymess/backend/dialer/__init__.py:
```diff
@@ -4,83 +4,67 @@
+    """Controller class for dialer delegating message to correct dialer backend"""
 
     model = DialerMessage
+    backend_type_name = 'DIALER'
```
could be. Changed

jsilhan commented on pymess/backend/__init__.py:
```diff
@@ -1,27 +1,156 @@
 
 
-class BaseBackend:
+class BaseController:
```
added

jsilhan commented:
@matllubos @rubickcz comments have been addressed

jsilhan commented:
@matllubos rebased
